### ISSUE TITLE
c++ virtual functions not working

### ISSUE THREAD

**Issue by vertexpotato** [emscripten-core/emscripten#16333]
I'm trying to use virtual functions with emscripten like the following : 

```
IGameAssembly* asm =  Application::GetGameAssembly();
auto scripts = asm ->GetScripts();

for (auto it : scripts)
{
       std::cout << it << std::endl;
}
```

the above doesn't work when calling the class interface pointer

but if I do it like this it works : 

```
WebGameAssembly* asm =  (WebGameAssembly*)Application::GetGameAssembly();
auto scripts = asm->GetScripts();

for (auto it : scripts)
{
       std::cout << it << std::endl;
}
```

any idea on how I can debug this ?

**Comment by sbc100** [comment]
In general there is no problem with virtual functions.. we have a huge about the continuous testing so for sure it works in the general case.

Do debug you specific I would suggest building with `-g` and debugging with chrome's devtools (https://developer.chrome.com/blog/wasm-debugging-2020/).

Building with `-g` will also give you a full stack trace so you should be able to see roughly where the null pointer is coming from.

**Comment by sbc100** [comment]
It seems unlikely, although not impossible, that this is an emscripten specific issue.  Does the same C++ code when compiled with other compilers?

**Comment by sbc100** [comment]
Ah, I think it could relate to the fact that you are using dynamic linking + pthreads which is still experimental.  Do you need `-sMAIN_MODULE`?

Also, can you try again with the latest release from last night (3.1.5) since that includes some fixes related to dylink+pthreads.

**Comment by vertexpotato** [comment]
> Ah, I think it could relate to the fact that you are using dynamic linking + pthreads which is still experimental. Do you need `-sMAIN_MODULE`?
> 
> Also, can you try again with the latest release from last night (3.1.5) since that includes some fixes related to dylink+pthreads.

Yes I need dynamic linking because, unfortunately my engine is designed around dynamic linking.

sure I'll update my emsdk and try...

**Comment by sbc100** [comment]
How about pthreads?  As an experiment can you try building without threads to confirm that this is the issue (I'm pretty sure it is at this point ..but its good to be able to remove some of the flags when debugging issues like this).

**Comment by vertexpotato** [comment]
> How about pthreads? As an experiment can you try building without threads to confirm that this is the issue (I'm pretty sure it is at this point ..but its good to be able to remove some of the flags when debugging issues like this).

confirmed, it works with pthreads removed. however I need pthreads to use my application, what can I do to help find the issue ?

**Comment by sbc100** [comment]
First lets see if 3.1.5 fixes your issue. 

If that doesn't work we need to start looking how to build a repro case (preferably a small one) that you can share.. but hopefully 3.1.5 will be the fix.

A couple more questions:
1. Have you done any side module loading at the time when this error occurs? 
2. Do you use runtime dynamic linking (dlopen) or load-time dynamic linking, or both?

**Comment by vertexpotato** [comment]
> First lets see if 3.1.5 fixes your issue.
> 
> If that doesn't work we need to start looking how to build a repro case (preferably a small one) that you can share.. but hopefully 3.1.5 will be the fix.
> 
> A couple more questions:
> 
>     1. Have you done any side module loading at the time when this error occurs?
> 
>     2. Do you use runtime dynamic linking (dlopen) or load-time dynamic linking, or both?

Yeah, I load side modules at the time that the error occurs, and I used load-time dynamic linking
and I'm still waiting for the emsdk to finish updating, I'll try to recompile my project with it when it's done...

**Comment by sbc100** [comment]
Currently there are issue with sync'ing code between the different thread.

There is a function called `_emscripten_thread_sync_code` which you can call to bring the current thread up-to-date but we currently have not way to force this to run on all threads.   Can you try calling this function before the crashing code?   If that fixes the issue you will want to looking into finding some way to have all your threads call this after each dlopen().

Even with this `_emscripten_thread_sync_code` being called I believe there could still be issues wasm table not being in sync between threads.

**Comment by vertexpotato** [comment]
> Currently there are issue with sync'ing code between the different thread.
> 
> There is a function called `_emscripten_thread_sync_code` which you can call to bring the current thread up-to-date but we currently have not way to force this to run on all threads. Can you try calling this function before the crashing code? If that fixes the issue you will want to looking into finding some way to have all your threads call this after each dlopen().
> 
> Even with this `_emscripten_thread_sync_code` being called I believe there could still be issues wasm table not being in sync between threads. This is feature that we are still working on

I have tried 3.1.5, it didn't fix the issue, however it gave a different error message ```RuntimeError: table index is out of bounds``` when it tries to call the virtual function.

I also tried calling ```_emscripten_thread_sync_code``` and it gave the same error message above.

**Comment by vertexpotato** [comment]
> Currently there are issue with sync'ing code between the different thread.
> 
> There is a function called `_emscripten_thread_sync_code` which you can call to bring the current thread up-to-date but we currently have not way to force this to run on all threads. Can you try calling this function before the crashing code? If that fixes the issue you will want to looking into finding some way to have all your threads call this after each dlopen().
> 
> Even with this `_emscripten_thread_sync_code` being called I believe there could still be issues wasm table not being in sync between threads. This is feature that we are still working on

@sbc100 
When stepping into ```_emscripten_thread_sync_code ``` the thread appears to be already in sync so i
![sync](https://user-images.githubusercontent.com/18310807/154864068-d375b176-59e9-4594-9108-98200c572de6.PNG)
t just returns...

**Comment by vertexpotato** [comment]
I have reached a dead end here, any workaround would be appreciated. even if I have to fork the entire compiler....

**Comment by sbc100** [comment]
Since we don't know the exact problem its hard to suggest a workaround.   The most useful thing would be repro case .. the smaller the better.

**Comment by royalstream** [comment]
I was having the same issue (not anymore).
I was trying to use a C++ library but the virtual functions kept failing with this error message `RuntimeError: table index is out of bounds`. I switched to the C-only version of the library just to fail at their hand-made implementation of vtables.
After lots of research I found a block of memory that was being accessed after being freed.

Apparently, virtual functions and vtables are the first ones to suffer when you have memory issues.

**Comment by Splizard** [comment]
@royalstream I've also run into an issue with calling virtual function pointers specifically, resulting in `table index is out of bounds`. This does not occur when calling non-virtual functions.
